### ISSUE TITLE
OAuth, persistent storage, and support for fallback authentication

### ISSUE THREAD

**Issue by bjompen** [AZDOPS/AZDOPS#176]
We should change the way we authenticate. 
Current plan is this:

- Take a dependency on module [AzAuth](https://github.com/PalmEmanuel/AzAuth) to easier support checking out OAuth tokens
- If no OAuth token is found, pop up an interactive login - OR- 
- Use PAT if it is located inside module scoped variable just the way it is currently working.

This would allow us to use the logged in sessions from az cli, `az login`, and Azure PowerShell, `Connect-AzAccount`
As such it would also allow ADOPS to use the recently added support for managed identities.

Since az authentication is cross session it would also help us "automatically" get support for persistent tokens. One login would be sufficient for the lifetime of a token.

One issue I see with this so far is that some endpoints may still lack support for OAuth. If we find one of those my suggestion is to implement a custom decorator class for any commands that should default to using PAT (something like `[EndpointRequiresPat()]`) to be able to enforce pat for this specific endpoint.

The only bigger issue I've identified is that OAuth tokens only seem to be one login at a time. This means  no more support for multi org authentication, unless we do something like 
```PowerShell
if ($Organization.Default) {
  # Use Oauth
}
else {
  # Use PAT or force a new OAuth login.
}
```

I'm not certain using pat here is a good idea as it might lead to uncertainty on what kind of auth is currently beeing used,
but forcing a new login may result in stopped pipelines.. Ideas are welcome.

**Comment by bjompen** [comment]
Thoughts:
- wrap Get-AzToken with the correct id for DevOps to make it easier to use, get-adopsoauthtoken
- getADOPSHeader 
  - do we have a cached OAuth token?
  - if no, can we get and cache one
  - if no, do we have a pat token?
  - if no, throw login error
- connect-ADOPS 
  - add parameter set OAuth that takes token output from get-adopsoauthtoken
- Internal helper function GetOrCacheOauthToken or similar

**Comment by SimonWahlin** [comment]
I think we can close this now, it was implemented in #189
